morebits.js

```diff
@@ -1152,6 +1152,52 @@ Morebits.array = {
 	}
 };
 
+/**
+ * select2 doesn't natively support highlightening of search hits so we implement that here
+ * @param {jQuery} $select
+ * @param {boolean} [hasOptionGroups=false] - does the select menu have optgroups?
+ * @requires select2
+ */
+Morebits.select2SearchHighlights = function select2SearchHighlights($select, hasOptionGroups) {
+	$select.on('select2:open', function() {
+		$('.select2-search__field')[0].addEventListener('keyup', function() {
+			var $ul = $('.select2-results__options');
+			$ul.find('.search-hit').each(function(_, e) {
+				var li_element = e.parentElement;
+				// This would convert <li>Hello <span class=search-hit>wo</span>rld</li>
+				// to <label>Hello world</label>
+				li_element.innerHTML = li_element.textContent;
+			});
+
+			if (this.value) {
+				var searchString = this.value;
+				var searchRegex = new RegExp(mw.RegExp.escape(searchString), 'i');
+
+				var selectorString;
+				// the odd way in which select2 organises the ul necessitates this hack
+				if (hasOptionGroups) {
+					selectorString = 'li > strong, li > ul > li';
+				} else {
+					selectorString = 'li';
+				}
+				$ul.find(selectorString).each(function() {
```

Comment:
There's a weird bug here.  Try this:

1. Open the warn menu, select single issue (or notice or custom or all templates)
2. Open the search menu; clicking is sufficient, even without search
3. Select level 1.  Try to search for something.

When you type, you get:
`VM765:1190 Uncaught DOMException: Failed to execute 'setStart' on 'Range': There is no child at offset`...

You can also trigger `setEnd`.

My initial guess was that the results here in `$ul` weren't being properly destroyed or recreated so there are counting issues, but it only issues the errors when on a menu with `optgroup`s after visiting any menu without `optgroup`s.  That would suggest otherwise, something to do with the `selectorString`?

Reply:
Funny. The searching works fine, but the console is flooded with errors. 😑

Reply:
Fixed.

Reply:
It amazes me how you unearth such obscure and well-hidden bugs, even when they leave no visual trace. Thanks a lot for finding this one. What was happening is that multiple event listeners were building up on the $select (1 added every time the main_group is changed), so the entire highlighting code was executing multiple times. In one of these listeners, `hasOptionGroups` is false and hence the wrong `selectorString` is used - resulting in errors. But the searching itself worked correctly because there were other concurrent event listeners doing the right job!